## Importamos librerias necesarias para iniciar EDA

In [52]:
import pandas as pd
import numpy as np
import warnings
import re
import json
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

# Text
import nltk
nltk.download('punkt')
nltk.download('stopwords')

from nltk import regexp_tokenize


warnings.filterwarnings('ignore')#to filter all the warnings
pd.set_option('float_format', '{:.4f}'.format)# to keep the float values short

[nltk_data] Downloading package punkt to /home/kennydev/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /home/kennydev/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## Obtenemos la data necesaria

1. Youtube Trending Statistics `(CSV)` - *Todo*

In [53]:
df_youtube = pd.read_csv('../data/raw/USvideos.csv')
df_youtube.head()

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

2. Obtenemos la información contenida en el archivo `JSON
`

In [54]:
with open('../data/raw/US_category_id.json', 'r') as f:
    category_us = json.loads(f.read())

df_us_cat = pd.json_normalize(category_us, record_path='items')
df_us_cat.rename({'id':'category_id'}, axis=1, inplace=True)
df_us_cat['category_id'] = df_us_cat['category_id'].astype('int64')
df_us_cat.head(3)

kind                                               etag  \
0  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/Xy1mB4_yLrHy_BmKm...   
1  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/UZ1oLIIz2dxIhO45Z...   
2  youtube#videoCategory  "m2yskBQFythfE4irbTIeOgYYfBU/nqRIq97-xe5XRZTxb...   

   category_id         snippet.channelId     snippet.title  snippet.assignable  
0            1  UCBR8-60-B28hp2BmDPdntcQ  Film & Animation                True  
1            2  UCBR8-60-B28hp2BmDPdntcQ  Autos & Vehicles                True  
2           10  UCBR8-60-B28hp2BmDPdntcQ             Music                True

3. Para tener toda la información en un solo `DataFrame` realizamos un merge con los datasets creados anteriormente **df_us_cat** y **df_youtube** 

In [55]:
df = pd.merge(df_youtube, df_us_cat, how='left', on='category_id')
df.head(1)

video_id trending_date                               title  \
0  2kyS6SvSYSE      17.14.11  WE WANT TO TALK ABOUT OUR MARRIAGE   

  channel_title  category_id              publish_time             tags  \
0  CaseyNeistat           22  2017-11-13T17:13:01.000Z  SHANtell martin   

    views  likes  dislikes  ...  \
0  748374  57527      2966  ...   

                                   thumbnail_link comments_disabled  \
0  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg             False   

   ratings_disabled  video_error_or_removed  \
0             False                   False   

                                         description                   kind  \
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  youtube#videoCategory   

                                                etag  \
0  "m2yskBQFythfE4irbTIeOgYYfBU/xId8RX7vRN8rqkbYZ...   

          snippet.channelId   snippet.title snippet.assignable  
0  UCBR8-60-B28hp2BmDPdntcQ  People & Blogs               True  

[1 rows x 21 columns]

## Limpieza de los datos

1. Existen columnas que no aportan información relevante a nuestro EDA, algunas de ellas son:
'thumbnail_link','kind','etag','id','snippet.assignable','snippet.channelId','channelId'

* thumbnail_link : imagenes en miniatura
* kind : columna con texto irrelevante
* etad : igual que kind 

In [56]:
columns_drop = ['thumbnail_link','kind','etag','snippet.assignable','snippet.channelId']
df = df.drop(columns_drop, axis=1)
print("DataFrame post cleaned shape: {}".format(df.shape))

DataFrame post cleaned shape: (40949, 16)


2. Observemos el comportamiento y coherencia de los datos

**Premisas**
* Un video que no tenga visitas, no puede tener likes ni dislikes
* La cantidad de likes o dislike, no puede ser mayor a la cantidad de visitas
* Un video con las reacciones deshabilitadas, no puede tener like ni dislikes
* Un video con comentarios deshabilitados, su cantidad de comentarios debe ser cero

In [57]:
df.sample(3)

video_id trending_date  \
35477  1csFTDXXULY      18.18.05   
29859  dWLDoQ0Kus8      18.20.04   
4127   NqntFonA2iY      17.04.12   

                                                   title channel_title  \
35477  Can you solve the false positive riddle? - Ale...        TED-Ed   
29859                         We Got Coachella Makeovers         As/Is   
4127   Jennifer Lawrence: 'I Become Incredibly Rude' ...       Variety   

       category_id              publish_time  \
35477           27  2018-05-08T15:02:53.000Z   
29859           22  2018-04-12T00:17:18.000Z   
4127            24  2017-11-28T19:26:57.000Z   

                                                    tags   views  likes  \
35477  TED|"TED-Ed"|"Ted Education"|"Teded"|"TED Ed"|...  476386  12919   
29859  BuzzFeed|"BuzzFeed Video"|"Coachella"|"Fashion...  862336  21485   
4127   Variety|"Variety Studio"|"Actors on Actors"|"J...  638979   2964   

       dislikes  comment_count  comments_disabled  ratings_disabled  \
35477       416           1395              False             False   
29859      1028           1447              False             False   
4127       7613           3317              False             False   

       video_error_or_removed  \
35477                   False   
29859                   False   
4127                    False   

                                             description   snippet.title  
35477  Practice more problem-solving at https://brill...       Education  
29859  I feel like Beyonce, Zenon: Girl of the 21st C...  People & Blogs  
4127   Jennifer Lawrence and Adam Sandler sat down to...   Entertainment

In [58]:
# Video sin visitas, no tiene likes ni dislikes
df[(df['views'] == 0) & ((df['likes'] > 0) | (df['dislikes'] > 0))]

Empty DataFrame
Columns: [video_id, trending_date, title, channel_title, category_id, publish_time, tags, views, likes, dislikes, comment_count, comments_disabled, ratings_disabled, video_error_or_removed, description, snippet.title]
Index: []

In [59]:
df.describe()

category_id          views        likes     dislikes  comment_count
count   40949.0000     40949.0000   40949.0000   40949.0000     40949.0000
mean       19.9724   2360784.6383   74266.7024    3711.4009      8446.8037
std         7.5683   7394113.7597  228885.3382   29029.7059     37430.4870
min         1.0000       549.0000       0.0000       0.0000         0.0000
25%        17.0000    242329.0000    5424.0000     202.0000       614.0000
50%        24.0000    681861.0000   18091.0000     631.0000      1856.0000
75%        25.0000   1823157.0000   55417.0000    1938.0000      5755.0000
max        43.0000 225211923.0000 5613827.0000 1674420.0000   1361580.0000

No existen videos con un numero de vistas igual o menor a cero, lo que guarda relación con la cantidad de likes y dislikes

In [60]:
# Cantidad de likes o dislikes, no puede ser mayor a cantidad de views
df[(df['likes'] > df['views']) | (df['dislikes'] > df['views'])].sum()

video_id                 0.0000
trending_date            0.0000
title                    0.0000
channel_title            0.0000
category_id              0.0000
publish_time             0.0000
tags                     0.0000
views                    0.0000
likes                    0.0000
dislikes                 0.0000
comment_count            0.0000
comments_disabled        0.0000
ratings_disabled         0.0000
video_error_or_removed   0.0000
description              0.0000
snippet.title            0.0000
dtype: float64

No existen videos que cuenten con mayor cantidad de likes o dislikes que vistas

In [61]:
# Videos con reacciones deshabilitadas, no puede tener likes ni dislikes
df[(df['ratings_disabled'] == True) & (df['likes'] > 0)].sum()

video_id                 0.0000
trending_date            0.0000
title                    0.0000
channel_title            0.0000
category_id              0.0000
publish_time             0.0000
tags                     0.0000
views                    0.0000
likes                    0.0000
dislikes                 0.0000
comment_count            0.0000
comments_disabled        0.0000
ratings_disabled         0.0000
video_error_or_removed   0.0000
description              0.0000
snippet.title            0.0000
dtype: float64

In [62]:
df[(df['ratings_disabled'] == True) & (df['dislikes'] > 0)].sum()

video_id                 0.0000
trending_date            0.0000
title                    0.0000
channel_title            0.0000
category_id              0.0000
publish_time             0.0000
tags                     0.0000
views                    0.0000
likes                    0.0000
dislikes                 0.0000
comment_count            0.0000
comments_disabled        0.0000
ratings_disabled         0.0000
video_error_or_removed   0.0000
description              0.0000
snippet.title            0.0000
dtype: float64

Los videos con las reacciones deshabilitads, no cuentan con likes o dislikes

In [63]:
# Videos con comentarios deshabilitados, no deben contar con comentarios
df[(df['comment_count'] > 0) & (df['comments_disabled'] == True)].count()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
description               0
snippet.title             0
dtype: int64

Los videos con los comentarios deshabilitades, no cuentan con ningún comentario

<p style='color: green'>Conclusion</p>

La salud de las columnas númericas, y su coherencia parecen ser correctas, de acuerdo a las premisas analizadas anteriormente, por ello podemos hacer drop de las columnas `comments_disabled` y `ratings_disabled` ya que esta información puede ser interpretada de sus derivados `likes`, `dislikes` y `comment_count`

In [64]:
df = df.drop(['comments_disabled', 'ratings_disabled'], axis=1)

3. Analizaremos datos faltantes y duplicados en nuestro dataset

In [65]:
print("""
      ****************************************
            COLUMNAS CON DATOS NULOS
      ****************************************""")
print(df.isna().sum())




      ****************************************
            COLUMNAS CON DATOS NULOS
      ****************************************
video_id                    0
trending_date               0
title                       0
channel_title               0
category_id                 0
publish_time                0
tags                        0
views                       0
likes                       0
dislikes                    0
comment_count               0
video_error_or_removed      0
description               570
snippet.title               0
dtype: int64


Tenemos 570 registros `NaN` en columna descripcion, observaremos algunas filas con estos valores nulos para tomar una decisión sobre como lidiar con ellos

In [66]:
df[df['description'].isna()].sample(10)

video_id trending_date  \
9676   aE4XqQZLkek      18.01.01   
2568   k-qUBZWor4I      17.26.11   
4307   K7NE3nfdkgg      17.05.12   
2358   mN4JWnTOSmg      17.25.11   
20895  o21-6NnDmzY      18.28.02   
10128  S0eLwrrTBlc      18.03.01   
12507  HPs9EGBhl7M      18.17.01   
18531  Ld57wNXddGU      18.16.02   
10883  QB6TuFGuxj4      18.07.01   
35679  Hzk1bM2vVFU      18.19.05   

                                                   title  \
9676                              1982 TOYOTA COROLLA Ad   
2568   OU QB Baker Mayfield addresses the punishment ...   
4307                                   Creepy Matt lauer   
2358                   Drake - Virginia Black commercial   
20895                             OUR MOMENT. #MYCALVINS   
10128                    Oh What a Feeling. Toyota! 1982   
12507                          Pencil Sharpening Machine   
18531                         best pirate i've ever seen   
10883  Michael Jackson's Dont Stop Till You Get Enoug...   
35679                    True Facts : Carnivorous Plants   

              channel_title  category_id              publish_time  \
9676   automotivemaster1972           24  2017-09-15T16:28:32.000Z   
2568           Bob Przybylo           17  2017-11-21T00:24:05.000Z   
4307           Itzztonytony           23  2014-02-24T00:00:28.000Z   
2358        The Sports King           22  2017-11-19T20:23:56.000Z   
20895          Calvin Klein           24  2018-02-22T14:37:07.000Z   
10128          Kevin Noonan           27  2013-05-28T00:02:49.000Z   
12507      Stefany Gonzalez           22  2013-05-24T12:14:49.000Z   
18531                  hwow           20  2018-02-09T14:56:43.000Z   
10883                   JAY           24  2017-11-19T01:10:12.000Z   
35679              zefrank1           22  2018-05-07T19:50:01.000Z   

                                                    tags    views  likes  \
9676                                              [none]     2075      4   
2568                                              [none]    12904     18   
4307                                   Mattlauer|"Conan"   339306    528   
2358                                              [none]    23138    238   
20895                                             [none]   127543    819   
10128                                             [none]    13630     42   
12507                                             [none]    11630     40   
18531  best pirate i've ever seen|"best"|"pirate"|"i'...   185337   5003   
10883  Michael Jackson|"Dance"|"Dancing"|"Dancers"|"D...    55941    720   
35679                                             [none]  1425300  83921   

       dislikes  comment_count  video_error_or_removed description  \
9676          1              0                   False         NaN   
2568          7             51                   False         NaN   
4307         39             50                   False         NaN   
2358          5             38                   False         NaN   
20895       533            172                   False         NaN   
10128         1              8                   False         NaN   
12507         3             12                   False         NaN   
18531        57            186                   False         NaN   
10883        14             61                   False         NaN   
35679       765           5693                   False         NaN   

        snippet.title  
9676    Entertainment  
2568           Sports  
4307           Comedy  
2358   People & Blogs  
20895   Entertainment  
10128       Education  
12507  People & Blogs  
18531          Gaming  
10883   Entertainment  
35679  People & Blogs

No existe un patron definido, que nos indique la razón de la ausencia de descripciones en los videos del dataset, así que llenaremos estos valores nulos con `strings` vacios

In [67]:
df['description'].fillna('empty', inplace=True)

### Tags, Channel Title, Video Title y Descripciones con caracteres validos
* Strings que puedan ser decodificados bajo el protocolo ASCII - Columna `descriptions`

In [68]:
# Encuentra cadenas de caracteres que pueden o no ser decodificados por el protocolo ASCII
def is_english_string(text):
    try: 
        text.encode('ascii') 
    except UnicodeEncodeError: 
        return False
    else: 
        return True


In [69]:
# Remueve palabras con caracteres especiales y retorna una nueva cadena sin esas palabras

def remove_words_especial_char(text):
    new_text = []
    for word in text.split():
        if is_english_string(word):
            new_text.append(word)
    
    if len(new_text) <= 0:
        return None
    
    return " ".join(new_text)

In [70]:
df['des_is_english'] = df['description'].apply(lambda d: is_english_string(d))

In [71]:
df['cln_description'] = df['description'].apply(lambda d: remove_words_especial_char(d))

In [72]:
df[df['des_is_english'] == False][['cln_description', 'description']]

cln_description  \
2      WATCH MY PREVIOUS VIDEO \n\nSUBSCRIBE https://...   
3      Today we find out if Link is a Nickelback amat...   
5      Using the iPhone for the past two weeks -- her...   
7      Ice Cream Pint Combination Lock - http://amzn....   
12     WATCH SILENCE MUSIC VIDEO https://youtu.be/Tx1...   
...                                                  ...   
40935  Chun-Li (Live on SNL / 2018)Song Available Her...   
40938  Daddy Yankee - Hielo (Video Oficial)Spotify: h...   
40941  Stream + Download:https://wmna.sh/speechlessht...   
40946  I had so much fun transforming Safiyas hair in...   
40948  Call of Duty: Black Ops 4 Multiplayer raises t...   

                                             description  
2      WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3      Today we find out if Link is a Nickelback amat...  
5      Using the iPhone for the past two weeks -- her...  
7      Ice Cream Pint Combination Lock - http://amzn....  
12     WATCH SILENCE MUSIC VIDEO ▶ https://youtu.be/T...  
...                                                  ...  
40935  Chun-Li (Live on SNL / 2018)Song Available Her...  
40938  Daddy Yankee - Hielo (Video Oficial)Spotify: h...  
40941  Stream + Download:https://wmna.sh/speechlessht...  
40946  I had so much fun transforming Safiyas hair in...  
40948  Call of Duty: Black Ops 4 Multiplayer raises t...  

[23626 rows x 2 columns]

Terminado el procedimiento podemos hacer drop sobre las columnas temporales, creadas para esta limpieza

In [73]:
df.drop(['des_is_english', 'description'], axis=1, inplace=True)

* Strings que puedan ser decodificados bajo el protocolo ASCII - Columna `tags`

In [74]:
df['tags_is_english'] = df['tags'].apply(lambda t: is_english_string(t))

In [75]:
df['cln_tags'] = df['tags'].apply(lambda d: remove_words_especial_char(d))

In [76]:
df[df['tags_is_english'] == False][['cln_tags', 'tags']]

cln_tags  \
52     wwe|"world wrestling Top 10"|"Stone Cold Steve...   
82     cave|"cheese"|"goat"|"rind"|"sheep's milk"|"cr...   
88     Alan Walker|"DJ Walkzz"|"K-391"|"House"|"Techn...   
100    Japan|"Japanese"|"yt:cc=on"|"idiot's"|"guide"|...   
128    cartoon|"simons cat"|"simon's cat"|"simonscat"...   
...                                                  ...   
40874  Let's Lets Go"|"Pokemon Lets Go"|"Pokemon Let'...   
40887                                               None   
40893  cartoon|"simons cat"|"simon's cat"|"simonscat"...   
40896                                               None   
40923  camels|"cactus"|"pets"|"livestock"|"animals"|"...   

                                                    tags  
52     wwe|"world wrestling entertainment"|"wrestling...  
82     cave|"cheese"|"goat"|"rind"|"sheep's milk"|"cr...  
88     Alan Walker|"DJ Walkzz"|"K-391"|"House"|"Techn...  
100    Japan|"Japanese"|"yt:cc=on"|"idiot's"|"guide"|...  
128    cartoon|"simons cat"|"simon's cat"|"simonscat"...  
...                                                  ...  
40874  Pokémon|"Pokemon"|"Pokémon Let's Go"|"Pokémon ...  
40887  cook|"cat"|"Jun"|"Rachel"|"Japanese"|"food"|"j...  
40893  cartoon|"simons cat"|"simon's cat"|"simonscat"...  
40896  방탄소년단|"BTS"|"BANGTAN"|"HIPHOP"|"랩몬스터"|"RapMons...  
40923  camels|"cactus"|"pets"|"livestock"|"animals"|"...  

[2916 rows x 2 columns]

In [77]:
df.drop(['tags_is_english', 'tags'], axis=1, inplace=True)

* Strings que puedan ser decodificados bajo el protocolo ASCII - Columna `title`

In [78]:
df['title_is_english'] = df['title'].apply(lambda t: is_english_string(t))
df['cln_title'] = df['title'].apply(lambda d: remove_words_especial_char(d))
df[df['title_is_english'] == False][['cln_title', 'title']]

cln_title  \
9        Why the rise of the robots mean the end of work   
32                    Eminem - Walk On Water (Audio) ft.   
57     Kellyanne Conway on Roy Moore This Week Abc: N...   
68     Watch Norman Reedus Come Face to Face with his...   
75           Rosie On Donald Hostility Toward Her | WWHL   
...                                                  ...   
40874    Let's Go, Pikachu! and Let's Go, Eevee! Trailer   
40896      [CHOREOGRAPHY] BTS 'FAKE LOVE' Dance Practice   
40910           HONNE - Location Unknown (feat. Georgia)   
40923                               Camels vs. Cactus!!!   
40948  Official Call of Black Ops 4 Multiplayer Revea...   

                                                   title  
9      Why the rise of the robots won’t mean the end ...  
32            Eminem - Walk On Water (Audio) ft. Beyoncé  
57     Kellyanne Conway on Roy Moore This Week Abc: T...  
68     Watch Norman Reedus Come Face to Face with his...  
75     Rosie O’Donnell On Donald Trump’s Hostility To...  
...                                                  ...  
40874  Pokémon: Let's Go, Pikachu! and Pokémon: Let's...  
40896  [CHOREOGRAPHY] BTS (방탄소년단) 'FAKE LOVE' Dance P...  
40910         HONNE - Location Unknown ◐ (feat. Georgia)  
40923                         Camels vs. Cactus!!!   جمل  
40948  Official Call of Duty®: Black Ops 4 — Multipla...  

[3706 rows x 2 columns]

In [79]:
df.drop(['title_is_english', 'title'], axis=1, inplace=True)

* Strings que puedan ser decodificados bajo el protocolo ASCII - Columna `channel_title`

In [80]:
df['chnl_title_is_english'] = df['channel_title'].apply(lambda t: is_english_string(t))
df['cln_channel_title'] = df['channel_title'].apply(lambda d: remove_words_especial_char(d))
df[df['chnl_title_is_english'] == False][['cln_channel_title', 'channel_title']]

cln_channel_title                         channel_title
47           NBA Highlights YouTube              NBA Highlights · YouTube
82                              Bon                           Bon Appétit
296                         Lalonde                         Estée Lalonde
312          NBA Highlights YouTube              NBA Highlights · YouTube
359                             Bon                           Bon Appétit
...                             ...                                   ...
40464  The Official YouTube channel  The Official Pokémon YouTube channel
40615                           Bon                           Bon Appétit
40668  The Official YouTube channel  The Official Pokémon YouTube channel
40764                           Bon                           Bon Appétit
40874  The Official YouTube channel  The Official Pokémon YouTube channel

[652 rows x 2 columns]

In [81]:
df.drop(['chnl_title_is_english', 'channel_title'], axis=1, inplace=True)

Observemos nuestro dataset ahora

In [82]:
df.head(3)

video_id trending_date  category_id              publish_time    views  \
0  2kyS6SvSYSE      17.14.11           22  2017-11-13T17:13:01.000Z   748374   
1  1ZAPwfrtAFY      17.14.11           24  2017-11-13T07:30:00.000Z  2418783   
2  5qpjK5DgCt4      17.14.11           23  2017-11-12T19:05:24.000Z  3191434   

    likes  dislikes  comment_count  video_error_or_removed   snippet.title  \
0   57527      2966          15954                   False  People & Blogs   
1   97185      6146          12703                   False   Entertainment   
2  146033      5339           8181                   False          Comedy   

                                     cln_description  \
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...   
1  One year after the presidential election, John...   
2  WATCH MY PREVIOUS VIDEO \n\nSUBSCRIBE https://...   

                                            cln_tags  \
0                                    SHANtell martin   
1  last week tonight trump presidency|"last week ...   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   

                                           cln_title cln_channel_title  
0                 WE WANT TO TALK ABOUT OUR MARRIAGE      CaseyNeistat  
1  The Trump Presidency: Last Week Tonight with J...   LastWeekTonight  
2  Racist Superman | Rudy Mancuso, King Bach & Le...      Rudy Mancuso

## Reemplazo de caracteres especiales sobre columnas tipo object
* Deseamos dejar el modelo completamente preprocesado antes de crear cualquier modelo(optimizacion) u obtener insights

<p style='color: yellow'>Opciones</p>

* Usar la libreria `re` y con ella expresiones regulares
* Utilizar el modulo `str` incluido en python y con el la función `replace`


In [83]:
flex_pattern = r'''(?x)                  # Flag para iniciar el modo verbose
              (?:[A-Z]\.)+            # Hace match con abreviaciones como U.S.A.
              | \w+(?:-\w+)*         # Hace match con palabras que pueden tener un guión interno
              | \$?\d+(?:\.\d+)?%?  # Hace match con dinero o porcentajes como $15.5 o 100%
              | \.\.\.              # Hace match con puntos suspensivos
              | [][.,;'?"():-_`]    # Hace match con signos de puntuación
              
              '''

pattern = r'''(?x)                  # Flag para iniciar el modo verbose
              (?:[A-Z]\.)+            # Hace match con abreviaciones como U.S.A.
              | \w+(?:-\w+)*         # Hace match con palabras que pueden tener un guión interno
              | \$?\d+(?:\.\d+)?%?  # Hace match con dinero o porcentajes como $15.5 o 100%
              | \.\.\.              # Hace match con puntos suspensivos
              
              '''

In [89]:
def clean_words(text, pattern):
    if text == None:
        return None
    tokens = regexp_tokenize(text, pattern)
    return ' '.join(tokens)

1. Usando expresiones regulares sobre la columna `cln_tags`

In [90]:
df['post_cln_tags'] = df['cln_tags'].apply(lambda t: clean_words(t, pattern))
df['post_cln_tags'] = df['post_cln_tags'].apply(lambda t: t.lower() if t != None else None)
df[['post_cln_tags', 'cln_tags']].sample(10)

post_cln_tags  \
4771   shanedawsontv shane dawson vlogs iphone simila...   
9354   american impersonation impersonations of ameri...   
14044          kip moore last shot mca nashville country   
9111   thr the hollywood reporter hollywood reporter ...   
32800                                               hulu   
1355                           showtime shosports sports   
5505   merrell twins twins merrelltwins the merrell t...   
16723  foreigner speaking hindi haryanvi india smiles...   
3049   2017 animals cool cute dogs featured feel good...   
39949  horse racing horses horse racing triple crown ...   

                                                cln_tags  
4771   shanedawsontv|"shane"|"dawson"|"vlogs"|"iphone...  
9354   american impersonation|"impersonations of amer...  
14044  Kip|"Moore"|"Last"|"Shot"|"MCA"|"Nashville"|"C...  
9111   thr|"the hollywood reporter"|"hollywood report...  
32800                                               hulu  
1355                       showtime|"shosports"|"sports"  
5505   Merrell Twins|"Twins"|"Merrelltwins"|"The Merr...  
16723  Foreigner|"Speaking"|"Hindi"|"Haryanvi"|"India...  
3049   2017|"Animals"|"Cool"|"Cute"|"Dogs"|"Featured"...  
39949  Horse racing|"horses"|"horse"|"racing"|"Triple...

2. Usando expresiones regulares sobre la columna `cln_title`

In [92]:
df['post_cln_title'] = df['cln_title'].apply(lambda t: clean_words(t, pattern))
df['post_cln_title'] = df['post_cln_title'].apply(lambda t: t.lower() if t != None else None)
df[['post_cln_title', 'cln_title']].sample(10)

post_cln_title  \
35128                          season 4 announce trailer   
9754                  mamma mia here we go again trailer   
20912                                 haul video 2 tacos   
4481   making chocolate christmas pudding with mark f...   
3604       melania trump is dreaming of a dark christmas   
8919   ed sheeran s perfect scores christmas number 1...   
5601                                    ready player one   
1036   emergence how stupid things become smart together   
4738                 the michelangelo of microsoft excel   
40351  kardashian vs west let s meet the teams celebr...   

                                               cln_title  
35128                        SEASON 4 | ANNOUNCE TRAILER  
9754               Mamma Mia! Here We Go Again - Trailer  
20912                             Haul Video #2 | Tacos!  
4481   Making Chocolate Christmas Pudding with Mark F...  
3604       Melania Trump Is Dreaming Of A Dark Christmas  
8919   Ed Sheeran's Perfect scores Christmas Number 1...  
5601                                    ready player one  
1036   Emergence How Stupid Things Become Smart Together  
4738                 The Michelangelo of Microsoft Excel  
40351  Kardashian Vs West! Let's meet the teams! | Ce...

3. Usando expresiones regulares sobre la columna `cln_channel_title`

In [93]:
df['post_cln_channel_title'] = df['cln_channel_title'].apply(lambda t: clean_words(t, pattern))
df['post_cln_channel_title'] = df['post_cln_channel_title'].apply(lambda t: t.lower() if t != None else None)
df[['post_cln_channel_title', 'cln_channel_title']].sample(10)

post_cln_channel_title  \
24769                          hellthyjunkfood   
10678                              jackie aina   
36804                                    tasty   
36210                      movieclips trailers   
5615   watch what happens live with andy cohen   
2905                           e entertainment   
8032                               simon s cat   
17630                         calvinharrisvevo   
30206                    good mythical morning   
13760                           screen junkies   

                             cln_channel_title  
24769                          HellthyJunkFood  
10678                              Jackie Aina  
36804                                    Tasty  
36210                      Movieclips Trailers  
5615   Watch What Happens Live with Andy Cohen  
2905                          E! Entertainment  
8032                               Simon's Cat  
17630                         CalvinHarrisVEVO  
30206                    Good Mythical Morning  
13760                           Screen Junkies

In [ ]:
df.columns

Index(['video_id', 'trending_date', 'category_id', 'publish_time', 'views',
       'likes', 'dislikes', 'comment_count', 'video_error_or_removed',
       'snippet.title', 'cln_description', 'cln_tags', 'cln_title',
       'cln_channel_title', 'post_cln_tags', 'post_cln_title',
       'post_cln_channel_title'],
      dtype='object')

4. Usando expresiones regulares sobre la columna `cln_description`

In [94]:
df['post_cln_description'] = df['cln_description'].apply(lambda t: clean_words(t, pattern))
df['post_cln_description'] = df['post_cln_description'].apply(lambda t: t.lower() if t != None else None)
df[['post_cln_description', 'cln_description']].sample(10)

post_cln_description  \
24827  until the civil rights act passed in 1964 the ...   
32688  twice japan 3rd me 2018 05 16 me me up lemonad...   
19750  listen to i want u back from the sophomore alb...   
26846  enjoy the video subscribe here http bit ly 1fk...   
15416                                              empty   
34865  maluma disponible para pre-ordenar ahora itune...   
12720  this fair maiden has finally been escorted hom...   
35393                     jay https www facebook com jay   
14332  olivia munn and niecy nash were hilariously un...   
27764  to kevin hart not partner human coffee tables ...   

                                         cln_description  
24827  Until the Civil Rights Act passed in 1964, the...  
32688  TWICE JAPAN 3rd Me 2018.05.16 Me Me Up Lemonad...  
19750  Listen to I Want U Back from the sophomore alb...  
26846  Enjoy the video? Subscribe here: http://bit.ly...  
15416                                              empty  
34865  Maluma DISPONIBLE PARA PRE-ORDENAR AHORA!:iTun...  
12720  This fair maiden has finally been escorted hom...  
35393                   Jay https://www.facebook.com/jay  
14332  Olivia Munn and Niecy Nash were hilariously un...  
27764  To Kevin Hart, not partner human coffee tables...

Una vez terminado esta etapa, podemos eliminar las columnas que sufrieron el preprocesamiento y renombrar las persistentes

In [95]:
df.columns
columnas_repetidas = ['cln_description', 'cln_tags', 'cln_title', 'cln_channel_title', 'video_error_or_removed']
df.drop(columnas_repetidas, axis=1, inplace=True)

In [96]:
df.rename({'snippet.title':'category_name', 
           'post_cln_tags':'tags',
           'post_cln_title':'video_title',
           'post_cln_channel_title':'channel_title',
           'post_cln_description':'description'}, axis=1, inplace=True)
df.head(2)

video_id trending_date  category_id              publish_time    views  \
0  2kyS6SvSYSE      17.14.11           22  2017-11-13T17:13:01.000Z   748374   
1  1ZAPwfrtAFY      17.14.11           24  2017-11-13T07:30:00.000Z  2418783   

   likes  dislikes  comment_count   category_name  \
0  57527      2966          15954  People & Blogs   
1  97185      6146          12703   Entertainment   

                                                tags  \
0                                    shantell martin   
1  last week tonight trump presidency last week t...   

                                         video_title    channel_title  \
0                 we want to talk about our marriage     caseyneistat   
1  the trump presidency last week tonight with jo...  lastweektonight   

                                         description  
0  shantell s channel https www youtube com shant...  
1  one year after the presidential election john ...

In [97]:
df.tail(2)

video_id trending_date  category_id              publish_time  \
40947  oV0zkMe1K8s      18.14.06            1  2018-05-17T17:00:04.000Z   
40948  ooyjaVdt-jA      18.14.06           20  2018-05-17T17:09:38.000Z   

          views   likes  dislikes  comment_count     category_name  \
40947   5660813  192957      2846          13088  Film & Animation   
40948  10306119  357079    212976         144795            Gaming   

                                                    tags  \
40947  black panther hishe marvel infinity war how it...   
40948            call of duty cod activision black ops 4   

                                             video_title  \
40947                how black panther should have ended   
40948  official call of black ops 4 multiplayer revea...   

                  channel_title  \
40947  how it should have ended   
40948              call of duty   

                                             description  
40947  how black panther should have endedwatch more ...  
40948  call of duty black ops 4 multiplayer raises th...

## Tipos de datos adecuados

* Se detectaron columnas que contienen fechas con dtype object
* La columna `category_name` puede tener tipo `category`

In [98]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40949 entries, 0 to 40948
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   video_id       40949 non-null  object
 1   trending_date  40949 non-null  object
 2   category_id    40949 non-null  int64 
 3   publish_time   40949 non-null  object
 4   views          40949 non-null  int64 
 5   likes          40949 non-null  int64 
 6   dislikes       40949 non-null  int64 
 7   comment_count  40949 non-null  int64 
 8   category_name  40949 non-null  object
 9   tags           40628 non-null  object
 10  video_title    40927 non-null  object
 11  channel_title  40810 non-null  object
 12  description    40921 non-null  object
dtypes: int64(5), object(8)
memory usage: 4.4+ MB


In [99]:
def convert_str_to_datetime(string):
    date_time_obj = datetime.strptime(string, '%y.%d.%m')
    return date_time_obj

In [100]:
df['trending_date'] = df['trending_date'].apply(lambda date: convert_str_to_datetime(date))
df['publish_time'] = pd.to_datetime(df['publish_time']).dt.tz_convert(None)
df['category_name']= df['category_name'].astype('category')

In [101]:
df.head(2)

video_id trending_date  category_id        publish_time    views  likes  \
0  2kyS6SvSYSE    2017-11-14           22 2017-11-13 17:13:01   748374  57527   
1  1ZAPwfrtAFY    2017-11-14           24 2017-11-13 07:30:00  2418783  97185   

   dislikes  comment_count   category_name  \
0      2966          15954  People & Blogs   
1      6146          12703   Entertainment   

                                                tags  \
0                                    shantell martin   
1  last week tonight trump presidency last week t...   

                                         video_title    channel_title  \
0                 we want to talk about our marriage     caseyneistat   
1  the trump presidency last week tonight with jo...  lastweektonight   

                                         description  
0  shantell s channel https www youtube com shant...  
1  one year after the presidential election john ...

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40949 entries, 0 to 40948
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   video_id       40949 non-null  object        
 1   trending_date  40949 non-null  datetime64[ns]
 2   category_id    40949 non-null  int64         
 3   publish_time   40949 non-null  datetime64[ns]
 4   views          40949 non-null  int64         
 5   likes          40949 non-null  int64         
 6   dislikes       40949 non-null  int64         
 7   comment_count  40949 non-null  int64         
 8   category_name  40949 non-null  category      
 9   tags           40628 non-null  object        
 10  video_title    40927 non-null  object        
 11  channel_title  40810 non-null  object        
 12  description    40921 non-null  object        
dtypes: category(1), datetime64[ns](2), int64(5), object(5)
memory usage: 4.1+ MB


Hasta este punto parece que tenemos un Dataset limpio y listo para hacer nuestros primeros analisis, es momento de exportarlo, para ser usado cuando lo necesitemos.


In [103]:
df.to_csv('../data/preprocessed/video_cln.csv', index=False, na_rep=' ')

In [ ]:

# print("""
#       ****************************************
#       TOTAL DE DATOS DUPLICADOS EN EL DATASET
#       ****************************************""")
# print(df.duplicated().sum())